File: 백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Scanner;

public class Main {

    public static void bfs(int start, int end) {
        Deque<PositionNode> queue = new ArrayDeque<>();
        queue.add(new PositionNode(start, 0));

        int[] visited = new int[100001];
        for (int i = 0; i < visited.length; i++) {
            visited[i] = Integer.MAX_VALUE;
        }
        visited[start] = 0;

        int time = Integer.MAX_VALUE;
        int cnt = 0;

        while (!queue.isEmpty()) {
            PositionNode currentNode = queue.poll();
            int currentPos = currentNode.getPosition();
            int currentTime = currentNode.getTime();

            if (currentPos == end) {
                if (currentTime < time) {
                    time = currentTime;
                    cnt = 1;
                } else if (currentTime == time) {
                    cnt++;
                }
            }

            int[] array = {currentPos - 1, currentPos + 1, currentPos * 2};

            for (int position : array) {
                if (0 <= position && position <= 100000) {
                    if (currentTime + 1 <= visited[position]) {
                        visited[position] = currentTime + 1;
                        queue.add(new PositionNode(position, currentTime + 1));
                    }
                }
            }
        }

        System.out.println(time);
        System.out.println(cnt);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        String[] input = sc.nextLine().split(" ");
        int start = Integer.parseInt(input[0]);
        int end = Integer.parseInt(input[1]);
        bfs(start, end);
    }
}

class PositionNode {
    int position;
    int time;

    PositionNode(int position, int time) {
        this.position = position;
        this.time = time;
    }

    public int getPosition() {
        return this.position;
    }

    public int getTime() {
        return this.time;
    }
}

File: 백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.py
from collections import deque


def BFS(start, end):
    queue = deque()
    queue.append((start, 0))
    visited = {start: 0}
    time = float("inf")
    cnt = 0
    while queue:
        current_node, current_time = queue.popleft()
        if current_node == end:
            if current_time < time:
                time = current_time
                cnt = 1
            elif current_time == time:
                cnt += 1
            continue
        for next_node in (current_node - 1, current_node + 1, current_node * 2):
            if 0 <= next_node <= 100000:
                if next_node not in visited or current_time + 1 <= visited[next_node]:
                    visited[next_node] = current_time + 1
                    queue.append((next_node, current_time + 1))

    print(time)
    print(cnt)


if __name__ == "__main__":
    n, k = map(int, input().split())
    if n > k:
        print(n - k)
        print(1)
    else:
        BFS(n, k)
